File: codeforces/600-div2/e.py
def powr(p, x, s):
    # m = max(abs(p-x)-s, 0)
    # print("powr",p,x,s,m)
    return max(abs(p-x)-s, 0)


A = [(0, 0)]
inf = 9999
n, m = map(int, input().split())

for i in range(1, n+1):
    x, s = map(int, input().split())
    A.append((max(0, x-s), min(m, x+s)))
dp = [inf] * (m+1)
dp[0] = 0
A = sorted(A)

# for j in range(m+1):
#     dp[0][j] = 0
for p in range(1, m+1):

    for i in range(1, n+1):

        a, b = A[i]

        if p-1 >= a and p-1 <= b:
            dp[p] = dp[p-1]
        elif p > b:
            pwr = p-b +1
            dp[p] = min(dp[p], pwr + dp[p-pwr])


# for i in range(2, n+1):
#     x, s = A[i]
#     dp[i][1] = min(dp[i-1][1], powr(1, x, s))
#     for p in range(2, m+1):
#         dp[i][p] = max(dp[i][p-1], min(dp[i-1][p], powr(p, x, s)))


# print(A)
# for i in range(n+1):
    # print('\t'.join(str(dp[i][j]) for j in range(m+1)))

print(dp)

File: codeforces/600-div2/e.cpp
#include<stdio.h>

int min(int a,int b) {
	if (a>b) {
		return b;
	} else {
		return a;
	}
}

int max(int a,int b) {
	if (a<b) {
		return b;
	} else {
		return a;
	}
}

int main() {
	int n,m,x,s;
	scanf("%d%d",&n,&m);
	int A[m+1], B[m+1], dp[m+1];
	
	for (int i = 1; i<=n; i++) {
		scanf("%d%d", &x, &s);
		A[i] = x-s >0 ? x-s :0;
		B[i] = x+s >m ? m : x+s;
	}
	
	for (int p = 1; p<=m; p++) {
		dp[p] = 999999999;
	}
	dp[0] = 0;	
	for (int p = 1; p<=m; p++) {
		for (int i = 1; i<=n; i++) {
			int a = A[i];
			int b = B[i];
			if (p >= a && p <= b) {
				dp[p] = dp[p-1];				
			} else if (p < a) {
				int pwr = a - p;
				dp[p] = min(dp[p], min(pwr + dp[max(0,a-pwr-1)], p));
			} else if ( p > b) {
				int pwr = p - b;
				dp[p] = min(dp[p], min(pwr + dp[max(a-pwr-1,0)], p));
			}
		}
	}
	printf("%d",dp[m]);
	return 0;
}

